src/sliceDataStorage.cpp

```diff
@@ -469,4 +477,113 @@ bool SliceDataStorage::getExtruderPrimeBlobEnabled(int extruder_nr) const
+        if (smaller_support_islands.empty())
+        {
+            if (new_smaller_islands.empty())
+            {
+                // there is no islands to add, append this into the to_remove list
+                to_remove_indexes.push_back(part_idx);
+            }
+            else
+            {
+                // replace the current part with a newly generated part
+                support_infill_part.outline = new_smaller_islands.back();
+                new_smaller_islands.pop_back();
+            }
+            continue;
+        }
+
+        // there are one or more smaller parts.
+        // we first replace the current part with one of the smaller parts, for the rest, we keep replacing the
+        // parts that need to be removed until there is none, and then we put the rest into the newly generated
+        // parts list.
+        support_infill_part.outline = smaller_support_islands[0];
+
+        // reserve space beforehand
+        if (new_smaller_islands.capacity() < new_smaller_islands.size() + smaller_support_islands.size())
+        {
+            new_smaller_islands.reserve(new_smaller_islands.size() + smaller_support_islands.size());
+        }
+
+        for (size_t support_island_idx = 1; support_island_idx < smaller_support_islands.size(); ++support_island_idx)
+        {
+            const PolygonsPart& smaller_island = smaller_support_islands[support_island_idx];
+
+            if (to_remove_indexes_front_idx == to_remove_indexes.size())
+            {
+                // there is no parts that need to be removed, so we just add this as a new part
+                new_smaller_islands.push_back(smaller_island);
+            }
+            else
+            {
+                // replace the part that needs to be removed (FIFO) with this new part
+                const size_t to_replace_idx = to_remove_indexes[to_remove_indexes_front_idx++];
+                SupportInfillPart& to_replace_part = this->support_infill_parts[to_replace_idx];
+                to_replace_part.outline = new_smaller_islands.back();
+                new_smaller_islands.pop_back();
+            }
+        }
+    }
+
+#ifdef DEBUG
+    if (!to_remove_indexes.empty())
+    {
+        assert(new_smaller_islands.empty() && "to_remove and new_islands cannot both be non-empty");
+    }
+    if (!new_smaller_islands.empty())
+    {
+        assert(to_remove_indexes.empty() && "to_remove and new_islands cannot both be non-empty");
+    }
+#endif // DEBUG
+
+    // remove the ones that need to be removed (LIFO)
+    while (!to_remove_indexes.empty())
+    {
+        for (size_t remove_idx_idx = to_remove_indexes.size() - 1; remove_idx_idx >= to_remove_indexes_front_idx; --remove_idx_idx)
+        {
+            const size_t remove_idx = to_remove_indexes[remove_idx_idx];
+            this->support_infill_parts.erase(this->support_infill_parts.begin() + remove_idx);
```

Comment:
The way you do the erase here still makes for O(n^2).
Please look at my proposed code.